Website day9 (#49)
scalacenter/scala-advent-of-code: docs/puzzles/day9.md

can have size 2, 3 or 4 depending whether the cell is on the edge of the grid or
not:

```scala
type Height = Int
case class Position(x: Int, y: Int)

case class Heightmap(width: Int, height: Int, data: Vector[Vector[Height]]):

  def apply(pos: Position): Height = data(pos.y)(pos.x)

  def neighborsOf(pos: Position): List[(Position, Height)] =
    val Position(x, y) = pos
    List(
      Option.when(x > 0)(Position(x - 1, y)),
      Option.when(x < width - 1)(Position(x + 1, y)),
      Option.when(y > 0)(Position(x, y - 1)),
      Option.when(y < height - 1)(Position(x, y + 1))
    ).flatMap(List.from)
     .map(pos => pos -> apply(pos))
```

Using this method I implement the `lowPointsPositions` method which iterates over
all the cells in the grid and for each of them, it checks if the value in it is
smaller than the values of adjacent cells:

```scala
  def lowPointsPositions: LazyList[Position] =
    LazyList.range(0, height).flatMap { y =>
      LazyList.range(0, width).map { x => 
        val pos = Position(x, y)
        (
          apply(pos),
          pos,
          this.neighborsOf(pos).map(_._2)
        )
      }
    }
    .collect {

>       case (value, poss, neighbors) if neighbors.forall(value < _) => 
>         poss

Reviewer: ```suggestion
      case (value, pos, neighbors) if neighbors.forall(value < _) => 
        pos
```

    }
```

The method `forall` on collections returns true only if the predicate it is
provided with holds true for all elements of the collection. It is similar to
the `exists` methods which returns true if at least one element in the
collection satisfies the provided predicate.
Finally, `collect` allows us to simplify chains of `map` and `filter`.

> You can have a look at the documentation of these methods [in the
> documentation](https://www.scala-lang.org/api/current/scala/collection/immutable/Iterable.html)

Reviewer: "at the documentation ... in the documentation" is redundant. Consider:
```suggestion
You can find more information about these methods [in the
documentation](https://www.scala-lang.org/api/current/scala/collection/immutable/Iterable.html)
```

We now have all the tools to solve part 1:

```scala
def part1(input: String): Int =
  val heightMap = Heightmap.fromString(input)

  heightMap.lowPointsPositions.map(heightMap(_) + 1).sum
end part1
```

<Solver puzzle="day9-part1"/>

## Solution of Part 2

In part 2 we have to create, for each low point, a basin: a group of grid cells
that the smoke traverses while flowing downward towards the low point.

> This is how we can solve part 2 after implementing the basin creation for a single low point:

Reviewer: Nice to show the big picture before the details :+1:

```scala
  val lowPoints = heightMap.lowPointsPositions
  val basins = lowPoints.map(basin(_, heightMap))

  basins
    .to(LazyList)
    .map(_.size)
    .sorted(Ordering[Int].reverse)
    .take(3)
    .product
```

> To build a basin from a low point I use very useful data structure provided in

Reviewer: ```suggestion
To build a basin from a low point I use a very useful data structure provided in
```
but perhaps "very useful" is not very useful in that sentence ;) Consider removing it.

the standard library:
[`Queue`](https://www.scala-lang.org/api/current/scala/collection/immutable/Queue.html)
which implements a first-in-first-out
([FIFO](https://en.wikipedia.org/wiki/Queue_(abstract_data_type))) data
structure.

Starting from the low point, I retrieve the neighbors of a cell and check if

> they should be part of the basin, in other words whether they do not contain the

Reviewer: "whether they do not" is confusing. Consider rephrasing so that the sentence in the "whether" is positive.

digit `9`. I also remember the coordinates of the cells I visited using a `Set`
which provides constant time `contains` checks.  As the basin grows, the cells
that form it are stored in the `basinAcc` accumulator.  As I continue to
retrieve neighbors of neighbors, I add the cells that still need to be processed
in the queue.

> The algorithm stops when no cells to visit are available:

Reviewer: ```suggestion
The algorithm stops when there are no more cells to visit:
```
?
